### ISSUE TITLE
DSN::to_string() produces invalid DSN string

### ISSUE THREAD

**Issue by JonasKruckenberg** [getsentry/sentry-rust#505]
### Environment

`sentry-rust => "0.27.0"`
`@sentry/browser => ^7.2.0"`

### Steps to Reproduce

1. Parse `DSN` from string
2. call `.to_string()` on dsn
3. parse dsn string again to verify parsing fails (in my case this was parsed by the JS SDK, but that should not matter)

### Expected Result

string representation:
"https://db7855f82cec4baca8a0d6ec8d8f5d88@o4503930427473920.ingest.sentry.io/4503931193851907"

### Actual Result

string representation 
"https://db7855f82cec4baca8a0d6ec8d8f5d88:@o4503930427473920.ingest.sentry.io/4503931193851907"
notice the superfluous colon.

**Comment by Swatinem** [comment]
https://www.ietf.org/rfc/rfc1738.txt section 3.1 talks about this, and username with empty password is valid according to that RFC.

Can you elaborate which API you are using to parse things that is failing?

**Comment by AbhiPrasad** [comment]
We also should be parsing this properly in the JS SDK as well: https://github.com/getsentry/sentry-javascript/blob/master/packages/utils/test/dsn.test.ts, we can add more test cases to validate though if you can let us know the steps you are taking @JonasKruckenberg

**Comment by JonasKruckenberg** [comment]
Alright, thanks for your fast response 😉 

So TLDR; the JS SDK rejects that DSN string when the password is empty. I tracked this down to this line
https://github.com/getsentry/sentry-javascript/blob/8e967c339903a2bb6328f1f20c4a50712f78ed17/packages/utils/src/dsn.ts#L39 where the regex fails to account for empty password strings. 

Adding a star after the password capture group should fix this issue:

```diff
- const DSN_REGEX = /^(?:(\w+):)\/\/(?:(\w+)(?::(\w+))?@)([\w.-]+)(?::(\d+))?\/(.+)/;
+ const DSN_REGEX = /^(?:(\w+):)\/\/(?:(\w+)(?::(\w+)*)?@)([\w.-]+)(?::(\d+))?\/(.+)/;
```

For reference this the full process that I am doing:

I have a Rust backend and a JS frontend (a Tauri app) and I want to avoid people having to configure their DSN in two different places, so I have my Rust backend parse the DSN like so:
```rust
let sentry_options = sentry::ClientOptions {
    dsn:
      "https://db7855f82cec4baca8a0d6ec8d8f5d88@o4503930427473920.ingest.sentry.io/4503931193851907"
        .into_dsn()
        .expect("failed to parse DSN"),
    release: sentry::release_name!(),
    ..Default::default()
  };
```
and then later down the line inject that DSN into my javascript like so:
```rust
let js_init_script = format!(
    "window.__SENTRY_DSN__ = JSON.parse({:?})", dsn.expect("A DSN must be configured").to_string()
  );
```
the JS sentry SDK is then simply initialised the same way the getting started guide says, with the only exception that the DSN is taken from that window global:
```javascript
Sentry.init({
    dsn: window.__SENTRY_DSN__
  });
```
that last code snippet errors out when the DSN string contains that empty password.

**Comment by AbhiPrasad** [comment]
@timfish might be interested in what you are doing with Tauri! He did some work with https://github.com/timfish/sentry-tauri

**Comment by AbhiPrasad** [comment]
@JonasKruckenberg sounds good! Mind reaching out on [Discord](https://discord.com/invite/ez5KZN7)? Probably easier to get other folks involved that way and have a shorter response back and forth.

**Comment by Swatinem** [comment]
Ah nice, so this turned into an improvement of the JS SDK.

I will close this issue then, since there seems to be nothing wrong with `to_string` from the Rust side.
